### PR TITLE
Loop specialization

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Enables loop specialization through a connect mechanism. 
Three main use cases:

A) Set loop directly on the trainer:

```python
class MyFitLoop(FitLoop):
    # this loop has no nested loops
    ...

trainer = Trainer()
trainer.fit_loop = MyFitLoop()
```

B) Connect child loops to the main loop (nesting):

```python
class MyFitLoop(FitLoop):
    # this loop defines subloops
    
    def connect(self, child0, child1):
        self.child0 = child0
        ...

trainer = Trainer()
fit_loop = MyFitLoop()
child_loop = MyChildLoop()
fit_loop.connect(child_loop)

trainer.fit_loop = fit_loop
```

C) Connect a custom loop to an existing Lightning loop:

```python
class CustomBatchLoop(TrainingBatchLoop):
    # this loop is a custom implementation of the Lightning built-in batch loop
    ...

trainer = Trainer()
my_batch_loop = CustomBatchLoop()
trainer.fit_loop.epoch_loop.connect(my_batch_loop)
```

Setting the trainer references is handled internally by Lightning. The user does NOT connect the trainer directly to the loop.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/fit_loop.py [hunk 1]
```diff
@@ -32,25 +32,18 @@ class FitLoop(Loop):
     Args:
         min_epochs: The minimum number of epochs
         max_epochs: The maximum number of epochs
-        min_steps: The minimum number of steps
-        max_steps: The maximum number of epoch
```
Comment (yifuwang) [inline pytorch_lightning/loops/fit_loop.py:36]: nit: epoch -> steps

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -343,14 +346,30 @@ def __init__(
         self.slurm_connector = SLURMConnector(self)
         self.tuner = Tuner(self)
 
-        self.fit_loop = FitLoop(min_epochs, max_epochs, min_steps, max_steps)
+        # .fit() loop
+        self.fit_loop = FitLoop(
+            min_epochs=(1 if (min_epochs is None and min_steps is None) else min_epochs),
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:351]: Neat !

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -343,14 +346,30 @@ def __init__(
         self.slurm_connector = SLURMConnector(self)
         self.tuner = Tuner(self)
 
-        self.fit_loop = FitLoop(min_epochs, max_epochs, min_steps, max_steps)
+        # .fit() loop
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:349]: Should we move this to a Trainer Hook ? So people can override it ?

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:349]:
```
class Trainer

    def __init__(self, ..., connectors: BaseConnector):

        self.loop_connector = find_loop_connector(connectors) or LoopConnector()
        self.loop_connector.on_trainer_init(trainer)


class LoopConnector:

    def __init__(self):

    def on_trainer_init(self, trainer):
        self.trainer = trainer

        self.connect_loops()

    def connect_loops(self):
        # .fit() loop
        self.fit_loop = FitLoop(
            min_epochs=(1 if (min_epochs is None and min_steps is None) else min_epochs),
            max_epochs=(1000 if (max_epochs is None and max_steps is None) else max_epochs),
        )
        training_epoch_loop = TrainingEpochLoop(min_steps, max_steps)
        training_batch_loop = TrainingBatchLoop()
        validation_epoch_loop = EvaluationLoop()
        training_epoch_loop.connect(trainer=self, batch_loop=training_batch_loop, val_loop=validation_epoch_loop)
        training_batch_loop.connect(trainer=self)
        validation_epoch_loop.connect(trainer=self)
        self.fit_loop.connect(trainer=self, epoch_loop=training_epoch_loop)

        # .validate() loop
        self.validation_loop = EvaluationLoop()
        self.validation_loop.connect(trainer=self)

        # .test() loop
        self.test_loop = EvaluationLoop()
        self.test_loop.connect(trainer=self)

        # .predict() loop
        self.predict_loop = PredictionLoop()
        self.predict_loop.connect(trainer=self)
```

**Review by tchaton — COMMENTED** [review]
Hey @ananthsub @awaelchli, what are your thoughts there ?

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:361]: I think the user dealing with a connector is a bit too much to ask when it comes to loop customization. But we could have such a loop connector internally. 

I've been looking at simplifying the way the user would link loops together:


```python
    # construct loops
    fit_loop = FitLoop()
    train_epoch_loop = TrainingEpochLoop(min_steps=0, max_steps=2)
    train_batch_loop = TrainingBatchLoop()
    val_loop = EvaluationLoop()

    # link loops
    train_epoch_loop.link(batch_loop=train_batch_loop, val_loop=val_loop)
    fit_loop.link(epoch_loop=train_epoch_loop)

    # connect fit loop to trainer
    trainer.fit_loop = fit_loop
```

`internally, trainer calls connect()`

**Comment by awaelchli** [comment]
@tchaton @carmocca updated this PR based on the discussions and #8362

File: pl_examples/loop_examples/simple_loop.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    @property
+    def global_step(self) -> int:
+        return self.iteration_count
+
+    @property
+    def batch_idx(self) -> int:
+        # required by progress bar
+        return self.iteration_count
+
+    @property
```
Comment (tchaton) [inline pl_examples/loop_examples/simple_loop.py:39]: The loop could do this automatically.

File: pl_examples/loop_examples/simple_loop.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        # hook
+        self.trainer.call_hook("on_train_batch_end", output, batch, self.iteration_count, dataloader_idx=0)
+        self.trainer.call_hook("on_batch_end")
+
+    def on_run_end(self) -> None:
+        self.trainer.call_hook("on_train_end")
+        self.trainer.accelerator.on_train_end()
+        self.trainer._running_stage = None
+
+    def _run_optimization(self, batch: Any, batch_idx: int, optimizer: Optimizer):
```
Comment (tchaton) [inline pl_examples/loop_examples/simple_loop.py:82]: Lightning should provide those already implemented.

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 6]
```diff
@@ -66,14 +65,9 @@ def predictions(self):
...
 
-    def connect(
-        self, trainer: "pl.Trainer", *args: Any, progress: Optional[EpochLoopProgress] = None, **kwargs: Any
-    ) -> None:
-        """Connects the loop with necessary arguments like the trainer"""
-        super().connect(trainer, *args, **kwargs)
-        if progress is not None:
-            self.progress = progress
-        self.epoch_loop.connect(trainer, progress=self.progress.epoch)
+    def connect(self, epoch_loop: EvaluationEpochLoop):
```
Comment (tchaton) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:68]: Neat !

File: pl_examples/loop_examples/example2.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    # instantiate the new loop
+    simple_loop = SimpleLoop(num_iterations=1000)
+
+    # replace the fit loop
+    # the trainer reference will be set internally
+    trainer.fit_loop = simple_loop
+
+    # fit using the new loop!
+    trainer.fit(model, train_dataloader=train_data)
```
Comment (tchaton) [inline pl_examples/loop_examples/example2.py:37]: Love it !

**Review by tchaton — COMMENTED** [review]
Great job ! Do you think it is possible to provide _run_optimization or is it too soon ?

**Review by tchaton — APPROVED** [review]
Looks neat !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 8]
```diff
@@ -47,8 +52,8 @@ def __init__(self, min_steps: int, max_steps: int):
         self.batch_progress = Progress()
         self.scheduler_progress = SchedulerProgress()
 
-        self.batch_loop = TrainingBatchLoop()
-        self.val_loop = loops.EvaluationLoop()
+        self.batch_loop = Optional[TrainingBatchLoop]
+        self.val_loop = Optional["loops.EvaluationLoop"]
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:56]: Optional on the right hand side??

Did you mean?

```suggestion
        self.batch_loop: Optional[TrainingBatchLoop] = None
        self.val_loop: Optional["loops.EvaluationLoop"] = None
```

Regardless, why make it Optional? Not optional has the advantage of not having to deal with the `None` checks

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 9]
```diff
@@ -69,11 +74,16 @@ def done(self) -> bool:
...
-        """Connects the loop with necessary arguments like the trainer"""
-        super().connect(trainer, *args, **kwargs)
-        self.batch_loop.connect(trainer)
-        self.val_loop.connect(trainer)
+    def connect(
+        self,
+        batch_loop: Optional[TrainingBatchLoop] = None,
+        val_loop: Optional["loops.EvaluationLoop"] = None,
+    ) -> None:
+        """Optionally connect a custom batch- or validation loop to this training epoch loop."""
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:82]:
```suggestion
        """Optionally connect a custom batch or validation loop to this training epoch loop."""
```

Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py]: why? maybe it's just in german and not in english, but I wanted to abbreviate word repetation

batch loop or validation loop
to
batch- or validation loop

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py]: > Regardless, why make it Optional? Not optional has the advantage of not having to deal with the None checks

yes but mypy apparently needs the annotation if we set it to None.
and we can't set a default loop here, as we can't deal with initialization beforehand.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 9 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py]: Thought it was a typo.
Not sure if that's correct in English.

But AFAIK what you want is the difference between:

"batch or validation loop" -> "the batch loop or the validation loop"
vs
"batch, or validation loop" -> the batch or the validation loop"

File: pl_examples/loop_examples/example1.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    train_epoch_loop = TrainingEpochLoop(min_steps=0, max_steps=2)
+    train_batch_loop = TrainingBatchLoop()
+    val_loop = EvaluationLoop()
+
+    # connect loops together
+    train_epoch_loop.connect(batch_loop=train_batch_loop, val_loop=val_loop)
+    fit_loop.connect(epoch_loop=train_epoch_loop)
+
+    # connect fit loop to trainer (main entry point for the call in trainer.fit())
+    trainer.fit_loop = fit_loop
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pl_examples/loop_examples/example1.py]: no we have not, but 
1) internally it would look the same (self.fit_loop = loop)
2) to enable use case C from above in the PR description, user has to access the attribute anyway through the trainer

open for discussion of course

[the remark this answers is not in the document]
Comment (carmocca) [reply pl_examples/loop_examples/example1.py]: Okay. Yeah it won't work in all cases (point 2) so let's not provide it.

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 9 above)
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py]: okay, let's keep it as an exercise for the reader
xD

**Review by SeanNaren — APPROVED** [review]
